### PR TITLE
Remove Foundation from CTLine

### LINKED ISSUE #2357: CTLine should be implemented via CFRuntimeBase (not via an NSObject)

### PR DESCRIPTION
**Pull request by msft-Jeyaram**
- Implementation of CTLine via runtime based
- Reduced/Removed Attributed string creation in loops
- Implemented CFAttribute functions to append and insert attributed string
- Implemented NSAttributes in terms of CFAttributes
- Fixed memory leaks
- Code improvements
- Increased perf

Fixes #2357,#2451

image

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
Nit: please reword to be in imperative mood and not end in a period.

File: Frameworks/CoreText/CTLine.mm [hunk 1]
```diff
@@ -22,25 +22,89 @@
...
 #import <CoreFoundation/CFString.h>
 
-#include <memory>
+#import <memory>
 #import <algorithm>
 #import <numeric>
 #import <cwchar>
 #import <vector>
 
+static inline CFRange __CFRangeIntersectionRange(CFRange a, CFRange b) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: should this be made a more public helper function?

Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: CFIndex can be negative, so you need to make sure that the lengths are >= 0L

File: Frameworks/CoreText/CTLine.mm [hunk 2]
```diff
@@ -22,25 +22,89 @@
...
+    CFIndex start = std::max(a.location, b.location);
+    CFIndex end = std::min(a.location + a.length, b.location + b.length);
+    if (end <= start) {
+        return CFRangeMake(0, 0);
+    }
+    return CFRangeMake(start, end - start);
+}
+
+static void __CFAtributedStringAppendAttributedString(CFMutableAttributedStringRef attString, CFAttributedStringRef appendString) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: rename first param to attrString

File: Frameworks/CoreText/CTLine.mm [hunk 3]
```diff
@@ -22,25 +22,89 @@
...
+    CFRange replaceRange = { CFAttributedStringGetLength(attString), 0 };
+    CFAttributedStringReplaceString(attString, replaceRange, CFAttributedStringGetString(appendString));
+
+    CFRange enumerationRange = { 0, CFAttributedStringGetLength(appendString) };
+    CFIndex index = enumerationRange.location;
+    CFIndex prevlen = CFAttributedStringGetLength(attString);
+    CFIndex lenDiff = 0;
+    CFRange currentRange;
+
+    while ((index >= enumerationRange.location) && index < (enumerationRange.location + enumerationRange.length)) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: add some comments about what's going on in this

File: Frameworks/CoreText/CTLine.mm [hunk 4]
```diff
@@ -22,25 +22,89 @@
...
+        CFIndex currentlen = CFAttributedStringGetLength(attString);
+        if (prevlen != currentlen) {
+            lenDiff = currentlen - prevlen;
+            prevlen = currentlen;
+        }
+
+        index += (currentRange.length + lenDiff);
+        enumerationRange.length += lenDiff;
+
+        if (currentRange.length == 0) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: 0L.

File: Frameworks/CoreText/CTLine.mm [hunk 5]
```diff
@@ -22,25 +22,89 @@
...
+            break;
+        }
+    }
+}
+
+static inline void __CFAtributedStringInsertAttributedStringOneAttribute(CFMutableAttributedStringRef attString,
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: Attributed*

Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: Add some comments about what these methods do / why we need them in here

File: Frameworks/CoreText/CTLine.mm [hunk 6]
```diff
@@ -22,25 +22,89 @@
...
+}
+
+static inline void __CFAtributedStringAppendAttributedStringOneAttribute(CFMutableAttributedStringRef attString,
+                                                                         CFStringRef string,
+                                                                         CFDictionaryRef attribs) {
+    __CFAtributedStringInsertAttributedStringOneAttribute(attString, CFAttributedStringGetLength(attString), string, attribs);
+}
+
 static CFStringRef __CTCreateReversedString(CFStringRef string) {
     if (string == nullptr) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: RETURN_RESULT_IF...

File: Frameworks/CoreText/CTLine.mm [hunk 7]
```diff
@@ -22,25 +22,89 @@
...
         return nullptr;
     }
 
     CFIndex length = CFStringGetLength(string);
-    CFIndex usedBufLen;
-    CFStringGetBytes(string, CFRangeMake(0, length), kCFStringEncodingUTF16, 0, false, nullptr, length, &usedBufLen);
-
     if (length < 2) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: RETURN_RESULT_IF...

File: Frameworks/CoreText/CTLine.mm [hunk 8]
```diff
@@ -249,19 +276,17 @@ CTLineRef CTLineCreateJustifiedLine(CTLineRef line, CGFloat justificationFactor,
 /**
  @Status Interoperable
 */
-void CTLineDraw(CTLineRef lineRef, CGContextRef ctx) {
-    if (lineRef == nil || ctx == nil) {
-        return;
-    }
+void CTLineDraw(CTLineRef line, CGContextRef ctx) {
+    RETURN_IF(!line);
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: collapse into one RETURN_IF to match other methods

File: Frameworks/CoreText/CTLine.mm [hunk 9]
```diff
@@ -278,21 +303,24 @@ void CTLineDraw(CTLineRef lineRef, CGContextRef ctx) {
  @Status Interoperable
 */
 CFIndex CTLineGetGlyphCount(CTLineRef line) {
-    return line ? static_cast<_CTLine*>(line)->_glyphCount : 0;
+    RETURN_RESULT_IF_NULL(line, 0);
+    return line->_glyphCount;
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: These should all be getter methods

File: Frameworks/CoreText/CTLine.mm [hunk 10]
```diff
@@ -322,8 +347,7 @@ CGRect CTLineGetImageBounds(CTLineRef line, CGContextRef context) {
  @Status Interoperable
 */
 double CTLineGetTypographicBounds(CTLineRef lineRef, CGFloat* ascent, CGFloat* descent, CGFloat* leading) {
-    _CTLine* line = static_cast<_CTLine*>(lineRef);
-
+    __CTLine* line = const_cast<__CTLine*>(lineRef);
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: Just move this into a CTLine::GetTypographicBounds(...) so it doesn't need to be mutated like this.  Since the values are computed lazily just make them `mutable` so it can stay const.

File: Frameworks/CoreText/DWriteWrapper_CoreText.mm [hunk 11]
```diff
@@ -393,14 +393,14 @@ HRESULT STDMETHODCALLTYPE GetPixelsPerDip(_In_opt_ void* clientDrawingContext, _
  *
  * @return Unmutable array of _CTLine objects created with the requested parameters.
  */
-static _CTLine* _DWriteGetLine(CFAttributedStringRef string) {
+static CTLineRef _DWriteGetLine(CFAttributedStringRef string) {
```
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper_CoreText.mm]: I'm not sure why this one is Get since it returns a +1, but can you change this to Create*

File: include/CoreText/CoreTextExport.h [hunk 12]
```diff
@@ -1,6 +1,6 @@
 //******************************************************************************
 //
-// Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (aballway) [inline include/CoreText/CoreTextExport.h]: These are nice but again unrelated to the rest of these changes.  Pull these into a separate commit?

File: Frameworks/CoreText/CTLine.mm [hunk 13]
```diff
@@ -56,64 +120,95 @@ static CFStringRef __CTCreateReversedString(CFStringRef string) {
...
-    ret->_glyphCount = _glyphCount;
-    ret->_runs.attach(CFArrayCreateMutableCopy(kCFAllocatorDefault, 0, _runs));
-    ret->_relativeXOffset = _relativeXOffset;
+static CFMutableAttributedStringRef __getTruncatedStringFromSourceLine(CTLineRef sourceLine,
+                                                                       CTLineTruncationType truncationType,
+                                                                       double widthToExtract) {
+    CFArrayRef glyphRuns = CTLineGetGlyphRuns(sourceLine);
+    CFIndex numberOfRuns = CFArrayGetCount(glyphRuns);
+    double extractedWidth = 0;
+    bool done = 0;
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: false

File: Frameworks/CoreText/CTLine.mm (hunk 3 above)
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: just use CFAttributedStringReplaceAttributedString instead

File: Frameworks/CoreText/CTTypesetter.mm [hunk 14]
```diff
@@ -100,12 +98,13 @@ CFIndex CTTypesetterSuggestLineBreak(CTTypesetterRef typesetter, CFIndex startIn
  @Status Interoperable
 */
 CFIndex CTTypesetterSuggestLineBreakWithOffset(CTTypesetterRef typesetter, CFIndex index, double width, double offset) {
-      CTFrameRef frame = _DWriteGetFrame(typesetter->AttributedString(),
-                                      CFRangeMake(index, CFAttributedStringGetLength(typesetter->AttributedString()) - index),
-                                      CGRectMake(offset, 0, width, FLT_MAX));
+    CTFrameRef frame = _DWriteGetFrame(typesetter->AttributedString(),
+                                       CFRangeMake(index, CFAttributedStringGetLength(typesetter->AttributedString()) - index),
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTTypesetter.mm]: LOL

File: Frameworks/CoreText/CTLine.mm (hunk 13 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: LOL this is from earlier code.
I'll update it.

---
In reply to: [112725988](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112725988) [](ancestors = 112725988)

File: Frameworks/CoreText/CTLine.mm (hunk 3 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: this reflects what NSAttributedString is doing.
Given i don't have the full range of context on the CoreText dancing here, I decide to do the same.
which is also pretty trivial. If you are 100% this is sufficient, i can move forward. 

---
In reply to: [112734332](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112734332) [](ancestors = 112734332)

File: Frameworks/CoreText/CTLine.mm (hunk 7 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: LOL

---
In reply to: [112724123](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112724123) [](ancestors = 112724123)

File: Frameworks/CoreText/CTLine.mm (hunk 9 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: It would be ideal, but the earlier code violated the getter/setter pattern and exposed private vars outside the file.
so it doesn't help us much here.

---
In reply to: [112724807](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112724807) [](ancestors = 112724807)

File: Frameworks/CoreText/CTLine.mm (hunk 10 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: i'm not going to make logic changes in this CR.
That's out of scope, it should of been done initially to avoid modifying private vars.

---
In reply to: [112725004](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112725004) [](ancestors = 112725004)

File: Frameworks/CoreText/DWriteWrapper_CoreText.mm (hunk 11 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/DWriteWrapper_CoreText.mm]: I'm not sure why it was named like that too ;P


---
In reply to: [112725508](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112725508) [](ancestors = 112725508)

File: include/CoreText/CoreTextExport.h (hunk 12 above)
Comment (msft-Jeyaram) [inline include/CoreText/CoreTextExport.h]: The can easily be separated into a commit. 

---
In reply to: [112725826](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112725826) [](ancestors = 112725826)

File: Frameworks/CoreText/CTLine.mm (hunk 9 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: I mean in the paradigm this was done in CTLineGetGlyphCount _was_ the getter, but now since we're updating this to use CppBase and touching everything we should use better style 🛩

File: Frameworks/CoreText/CTLine.mm (hunk 10 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: It wouldn't be a logic change, just moving it into __CTLine like we do with other CppBase classes

Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: ah okay , you want to move the whole function into __CTLine.
uh, this is where it's going to get weird. __CTline is already violated elsewhere, so it has to go public strut.
Then this moves into it and expose a function. it would be ideal to clean everything up, so please open a bug for it.
The main purpose of this is to get rid of foundation.

---
In reply to: [112742309](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112742309) [](ancestors = 112742309)

File: Frameworks/CoreText/CTLine.mm (hunk 3 above)
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: Definitely just use CFAttributedStringReplaceAttributedString - I am 100% sure it is sufficient. The reason we don't use it for [NSMutableAttributedString appendAttributedString:] is to reduce the number of abstract functions. Before we made that class more properly abstract, the initial implementation of it DID use ReplaceAttributedString.

File: Frameworks/CoreText/DWriteWrapper_CoreText.mm (hunk 11 above)
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper_CoreText.mm]: @aballway didn't you name this

File: Frameworks/CoreText/CTLine.mm (hunk 10 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: This wouldn't require the rest of __CTLine be moved into a private struct (which would be better but would be some more churn).  We're already adding extra abstraction with _CTLineCreate so we might as well do this right 😸

Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: LOL yes it would not require for that to occur, uh but this just get's uglier.


---
In reply to: [112746377](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112746377) [](ancestors = 112746377)

File: Frameworks/CoreText/DWriteWrapper_CoreText.mm (hunk 11 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper_CoreText.mm]: Yup, this one is 100% my fault.  No clue why I named it that 🤷‍♂️

File: Frameworks/CoreText/CTLine.mm (hunk 9 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: inline functions in struct are pretty awesome ;P

---
In reply to: [112742078](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112742078) [](ancestors = 112742078)

**Comment by msft-Jeyaram** [comment]
@aballway  there is, #1 you aren't doing message dispatches anymore, second there some redundant code that was reduced + reduced allocations.
Updating this PR with run time perfs with benchmark.

**Comment by DHowett-MSFT** [comment]
Nit: please reword to be in imperative mood and not end in a period.

File: Frameworks/CoreText/CTLine.mm (hunk 1 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: sadly this guy has to go :(

---
In reply to: [112723252](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112723252) [](ancestors = 112723252)

File: Frameworks/CoreText/CTLine.mm (hunk 10 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: Also we need the cast anyways, since this method is modifying lineRef. so we can't do lineRef->GetTypographics...(), given GetTypographics can't be a const method.

---
In reply to: [112746739](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112746739) [](ancestors = 112746739,112746377)

**Comment by msft-Jeyaram** [comment]
@ms-jihua @DHowett-MSFT @aballway 
please review

File: Frameworks/CoreText/CTLine.mm [hunk 15]
```diff
@@ -28,79 +28,37 @@
...
+                                                                          CFStringRef string,
+                                                                          CFDictionaryRef attribs) {
     CFRange replaceRange = { index, 0 };
 
-    CFAttributedStringReplaceString(attString, replaceRange, string);
+    CFAttributedStringReplaceString(attrString, replaceRange, string);
 
     CFRange range = { 0, CFStringGetLength(string) };
-    CFAttributedStringSetAttributes(attString, { range.location + replaceRange.location, range.length }, attribs, true);
+    CFAttributedStringSetAttributes(attrString, { range.location + replaceRange.location, range.length }, attribs, true);
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: `range.location + replaceRange.location`

soooo...

`replaceRange.location`?

wait, sooo

`index` ?

File: Frameworks/CoreText/CTLine.mm (hunk 10 above)
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: The values being modified are for all intents and purposes constant for this method, they're just being lazily calculated, which is why they should be marked `mutable`.  I did something similar w/ CGDataProvider to keep things const even though the values were being lazily modified

File: Frameworks/include/CoreTextInternal.h [hunk 16]
```diff
@@ -48,16 +48,35 @@ inline void _SafeRelease(T** p) {
...
     CFRange _strRange;
     CGFloat _relativeXOffset;
     CGFloat _width;
-    NSUInteger _glyphCount;
+    CFIndex _glyphCount;
     woc::StrongCF<CFMutableArrayRef> _runs;
-    CGFloat _ascent, _descent, _leading;
-}
-@end
+    CGFloat _ascent;
```
Comment (aballway) [inline Frameworks/include/CoreTextInternal.h]: so then if these are marked `mutable` you'll be able to keep GetTypographicBounds() const 😄

File: Frameworks/CoreText/CTLine.mm [hunk 17]
```diff
@@ -306,50 +291,50 @@ CFRange CTLineGetStringRange(CTLineRef line) {
...
 }
 
-/**
- @Status Interoperable
-*/
-double CTLineGetTypographicBounds(CTLineRef lineRef, CGFloat* ascent, CGFloat* descent, CGFloat* leading) {
-    _CTLine* line = static_cast<_CTLine*>(lineRef);
-
-    RETURN_RESULT_IF((!line) || (CFArrayGetCount(line->_runs) == 0), 0);
+double __CTLine::GetTypographicBounds(CGFloat* ascent, CGFloat* descent, CGFloat* leading) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: <3

File: Frameworks/CoreText/CTLine.mm [hunk 18]
```diff
@@ -22,25 +22,47 @@
...
 #import <vector>
 
+// Adds the characters and attributes of a given attributed string to the end of the receiver.
+static inline void __CFAttributedStringAppendAttributedString(CFMutableAttributedStringRef attrString, CFAttributedStringRef appendString) {
+    CFRange replaceRange = { CFAttributedStringGetLength(attrString), 0 };
+    CFAttributedStringReplaceAttributedString(attrString, replaceRange, appendString);
+}
+
+// Inserts the characters and attribute of the given string into the receiver at the given index.
+static inline void __CFAttributedStringInsertAttributedStringOneAttribute(CFMutableAttributedStringRef attrString,
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: I don't think these would provide much perf improvement over your other non-OneAttribute helper, would they? I would like to just get rid of them.

File: Frameworks/include/CoreTextInternal.h (hunk 16 above)
Comment (msft-Jeyaram) [inline Frameworks/include/CoreTextInternal.h]: uh `mutable` 😄

---
In reply to: [112780387](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112780387) [](ancestors = 112780387)

File: Frameworks/CoreText/CTLine.mm (hunk 18 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: we keep on creating AttributedStrings and trying to insert them in the for-loop for based on the glyph count, it can be a bottleneck.
If glyph count is usually low, then this isn't needed.

---
In reply to: [112783999](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112783999) [](ancestors = 112783999)

**Comment by msft-Jeyaram** [comment]
image
